### ISSUE TITLE
Bound but stale PV are not reconciled

### ISSUE THREAD

**Issue by discordianfish** [hetznercloud/csi-driver#320]
Hi,

tl;dr:
A PV that is bound to a node that crashes hard doesn't recover due to the controller assuming that its still attached to the node.

At least I assume it's the case when the server crashes. In my case it's a bit less common setup: I'm running a single k3s node with it's state on a volume and a PV for some k8s app bound. I've re-created the node which came up healthy with all k8s state restored. But re-creating the pod that uses the PV failed since the csi driver assumed the disk is already attached to the server given that the PV is bound.

To fix that I'd suggest to check if the disk is really attached before trying to mount it.

**Comment by apricote** [comment]
Hello @discordianfish,

I think the problem you are describing happens because you re-use the same `Node` object in kubernetes for multiple servers. This causes Kubernetes to not reevaluate anything related to the Cloud-integrations, including the volume attachment.

>  A PV that is bound to a node that crashes hard doesn't recover due to the controller assuming that its still attached to the node.

Once Kubernetes removes the node and any pods, it will also cleanup the attached volume, see https://github.com/hetznercloud/csi-driver/issues/337#issuecomment-1352812843 for an explanation.

In your case, you can just delete the `VolumeAttachment` for the PV and Kubernetes should do the right thing. Or use a different recovery process that creates a new `Node` for the new server.

I am going to close the issue as this is outside of the csi-driver scope. Please feel free to reopen or create a new issue if you have any further questions or problems.
